sdk/src/lib.rs

```diff
@@ -145,8 +172,27 @@ where
                     return Poll::Pending;
                 }
 
-                // TODO: decode response.data to `D`.
-                Poll::Ready(Some(Ok(response)))
+                if response.message.is_none() {
+                    cx.waker().wake_by_ref();
+                    return Poll::Pending;
+                }
+
+                return match response.message.unwrap() {
+                    stream_data_response::Message::Data(m) => {
+                        let end_user_data = m
+                            .data
+                            .into_iter()
+                            .map(|b| D::decode(b.as_slice()))
+                            .filter_map(|b| b.ok())
+                            .collect::<Vec<D>>();
+                        return Poll::Ready(Some(Ok(end_user_data)));
+                    }
+                    // This case will handle Message::Heartbeat and Message::Invalidate
+                    _ => {
+                        cx.waker().wake_by_ref();
+                        Poll::Pending
+                    }
+                };
```

Comment:
Refactor to avoid unwrap. Also use the new `DataMessage` type.


```suggestion
                if response.stream_id != self.stream_id {
                    cx.waker().wake_by_ref();
                    return Poll::Pending;
                }

                match response.message {
                    None => {
                        cx.waker().wake_by_ref();
                        Poll::Pending
                    }
                    Some(stream_data_response::Message::Data(data)) => {
                        let batch = data
                            .data
                            .into_iter()
                            .map(|b| D::decode(b.as_slice()))
                            .filter_map(|b| b.ok())
                            .collect::<Vec<D>>();
                        let message = DataMessage::Data {
                            cursor: data.cursor,
                            end_cursor: data.end_cursor.unwrap_or_default(),
                            finality: DataFinality::from_i32(data.finality).unwrap_or_default(),
                            batch,
                        };
                        Poll::Ready(Some(Ok(message)))
                    }
                    Some(stream_data_response::Message::Invalidate(invalidate)) => {
                        let message = DataMessage::Invalidate {
                            cursor: invalidate.cursor,
                        };
                        Poll::Ready(Some(Ok(message)))
                    }
                    Some(stream_data_response::Message::Heartbeat(_)) => {
                        debug!("received heartbeat");
                        cx.waker().wake_by_ref();
                        Poll::Pending
                    }

```

Reply:
Done !